GET requests or bad payloads make the SDK panic

The receiver panics when receiving a GET request of empty payload.

Given this client
```go
func main() {
    cli, _ := cloudevents.NewDefaultClient()
    cli.StartReceiver(ctx, func(cloudevents.Event) {})
}
```

just `curl localhost:8080`.

<details><summary>📄 error trace</summary>

```
panic: runtime error: invalid memory address or nil pointer dereference
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x7b2946]

goroutine 19 [running]:
github.com/cloudevents/sdk-go/v2/protocol/http.(*Message).Finish(0x0, 0x0, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0-!r!c1/protocol/http/message.go:151 +0x26
github.com/cloudevents/sdk-go/v2/client.(*receiveInvoker).Invoke.func1(0xd91c20, 0x0, 0xc00006de78)
        /go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0-!r!c1/client/invoker.go:41 +0x47
panic(0xba6860, 0x128d010)
        /usr/local/go/src/runtime/panic.go:969 +0x166
github.com/cloudevents/sdk-go/v2/protocol/http.(*Message).ReadEncoding(0x0, 0xc000020000)
        /go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0-!r!c1/protocol/http/message.go:81 +0x5
github.com/cloudevents/sdk-go/v2/binding.ToEvent(0xd91c60, 0xc0000ae000, 0x7fce984681a0, 0x0, 0x0, 0x0, 0x0, 0xc00006ddbc, 0x2, 0x2)
        /go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0-!r!c1/binding/to_event.go:28 +0x52
github.com/cloudevents/sdk-go/v2/client.(*receiveInvoker).Invoke(0xc00009e560, 0xd91c60, 0xc0000ae000, 0xd91c20, 0x0, 0x0, 0x0, 0x0)
        /go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0-!r!c1/client/invoker.go:46 +0xfc
github.com/cloudevents/sdk-go/v2/client.(*ceClient).StartReceiver(0xc0000ce0a0, 0xd91c60, 0xc0000ae000, 0xb69e00, 0xcc3008, 0x0, 0x0)
        /go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0-!r!c1/client/client.go:218 +0x287
github.com/cloudevents/sdk-go/v2/client.(*obsClient).StartReceiver(0xc00009e540, 0xd91c60, 0xc0000ae000, 0xb69e00, 0xcc3008, 0x0, 0x0)
        /go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0-!r!c1/client/client_observed.go:93 +0xac
main.(*handler).Start.func1(0xc0000dc0c0, 0xc0000b03d0, 0xd91c60, 0xc0000ae000)
        /go/src/github.com/triggermesh/myproject/cmd/mycmdtarget/main.go:85 +0x63
created by main.(*handler).Start
        /go/src/github.com/triggermesh/myproject/cmd/mycmdtarget/main.go:84 +0x7a
exit status 2
```

</details>

The culprit is here:

https://github.com/cloudevents/sdk-go/blob/0fc16e13d87f0f2b1c99ec41183e35c0e68b6b5c/v2/binding/to_event.go#L23-L26

This comparison is `false` even when the request contains no payload.

---

edit: I observe the same result by sending valid payloads with a "bad" or missing `Content-Type` header:

```
$ curl -XPOST -H "Content-Type: application/json" localhost:8080/ -d'{
    "specversion" : "1.0",
    "type" : "com.example.someevent",
    "source" : "/mycontext",
    "id" : "A234-1234-1234",
    "time" : "2018-04-05T17:31:00Z",
    "comexampleextension1" : "value",
    "comexampleothervalue" : 5,
    "datacontenttype" : "text/xml",
    "data" : "<much wow=\"xml\"/>"
}'

// missing "+json" in the header -> boom
```

--- a/v2/client/client.go
+++ b/v2/client/client.go
@@ -216,7 +216,7 @@ func (c *ceClient) StartReceiver(ctx context.Context, fn interface{}) error {
 		}
 
 		if err != nil {
-			cecontext.LoggerFrom(ctx).Warn("Error while receiving a message: %s", err.Error())
+			cecontext.LoggerFrom(ctx).Warnf("Error while receiving a message: %s", err)
 			continue
 		}
 

--- a/v2/protocol/http/protocol.go
+++ b/v2/protocol/http/protocol.go
@@ -201,7 +201,12 @@ func (p *Protocol) Respond(ctx context.Context) (binding.Message, protocol.Respo
 		if !ok {
 			return nil, nil, io.EOF
 		}
+
+		if in.msg == nil {
+			return nil, in.respFn, in.err
+		}
 		return in.msg, in.respFn, in.err
+
 	case <-ctx.Done():
 		return nil, nil, io.EOF
 	}

--- a/v2/protocol/http/protocol_test.go
+++ b/v2/protocol/http/protocol_test.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"errors"
 	"net/http"
+	"net/http/httptest"
 	"testing"
 	"time"
 
@@ -11,6 +12,7 @@ import (
 	"github.com/cloudevents/sdk-go/v2/protocol"
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
@@ -48,13 +50,15 @@ func TestNew(t *testing.T) {
 }
 
 func protocols(t *testing.T) []*Protocol {
-	ps := make([]*Protocol, 0, 1)
+	ps := make([]*Protocol, 1)
 
 	p, err := New()
 	if err != nil {
-		t.Fail()
+		t.Fatalf("Failed to create test Protocol: %s", err)
 	}
-	ps = append(ps, p)
+
+	ps[0] = p
+
 	return ps
 }
 
@@ -133,14 +137,14 @@ func TestReceive(t *testing.T) {
 			wantErr: "nil Context",
 		},
 		"timeout": {
-			ctx:     context.TODO(),
+			ctx:     newDoneContext(),
 			wantErr: "EOF",
 		},
 	}
 	for n, tc := range testCases {
 		for _, p := range protocols(t) {
 			t.Run(n, func(t *testing.T) {
-				ReceiveTest(t, p, tc.ctx, tc.want, tc.wantErr)
+				ReceiveTest(t, p, tc.ctx, nil, tc.want, tc.wantErr)
 			})
 		}
 	}
@@ -161,27 +165,31 @@ func TestRespond(t *testing.T) {
 		"nil context": {
 			wantErr: "nil Context",
 		},
-		"timeout": {
-			ctx:     context.TODO(),
+		"timeout context": {
+			ctx:     newDoneContext(),
 			wantErr: "EOF",
+		},
+		"non-expiring context": {
+			ctx: context.Background(),
 		},
 	}
 	for n, tc := range testCases {
 		for _, p := range protocols(t) {
 			t.Run(n, func(t *testing.T) {
-				if tc.ctx != nil {
-					var done context.CancelFunc
-					tc.ctx, done = context.WithDeadline(tc.ctx, time.Now().Add(time.Millisecond*10))
-					defer done()
-				}
+				go func() {
+					time.Sleep(time.Millisecond * 10)
+					p.incoming <- msgErr{}
+				}()
 
 				got, fn, err := p.Respond(tc.ctx)
 				if tc.wantErr != "" {
-					if err == nil || err.Error() != tc.wantErr {
-						t.Fatalf("Expected error '%s'. Actual '%v'", tc.wantErr, err)
-					}
-				} else if err != nil {
-					t.Fatalf("Unexpected error: %v", err)
+					assert.EqualError(t, err, tc.wantErr)
+				} else {
+					assert.NoError(t, err)
+				}
+
+				if got != nil {
+					assert.NotNil(t, got, "Nil interface compares to nil")
 				}
 
 				if diff := cmp.Diff(tc.want, got); diff != "" {
@@ -213,6 +221,7 @@ func TestServeHTTP_Receive(t *testing.T) {
 		wantErr string
 	}{
 		"nil": {
+			rw:      httptest.NewRecorder(),
 			wantErr: "unknown Message encoding",
 		},
 	}
@@ -220,26 +229,24 @@ func TestServeHTTP_Receive(t *testing.T) {
 		for _, p := range protocols(t) {
 			t.Run(n, func(t *testing.T) {
 				go p.ServeHTTP(tc.rw, tc.req)
-				ReceiveTest(t, p, context.Background(), tc.want, tc.wantErr)
+				rec := (tc.rw).(*httptest.ResponseRecorder)
+				ReceiveTest(t, p, context.Background(), rec, tc.want, tc.wantErr)
 			})
 		}
 	}
 }
 
-func ReceiveTest(t *testing.T, p *Protocol, ctx context.Context, want binding.Message, wantErr string) {
-	if ctx != nil {
-		var done context.CancelFunc
-		ctx, done = context.WithTimeout(ctx, time.Millisecond*10)
-		defer done()
-	}
-
+func ReceiveTest(t *testing.T, p *Protocol, ctx context.Context, rec *httptest.ResponseRecorder, want binding.Message, wantErr string) {
 	got, err := p.Receive(ctx)
 	if wantErr != "" {
-		if err == nil || err.Error() != wantErr {
-			t.Fatalf("Expected error '%s'. Actual '%v'", wantErr, err)
+		assert.EqualError(t, err, wantErr)
+
+		if rec != nil {
+			defer rec.Result().Body.Close()
+			// TODO perform assertions on result if necessary
 		}
-	} else if err != nil {
-		t.Fatalf("Unexpected error: %v", err)
+	} else {
+		assert.NoError(t, err)
 	}
 
 	if want == nil {
@@ -263,4 +270,10 @@ func (r *roundTripperTest) RoundTrip(req *http.Request) (*http.Response, error) 
 	}
 
 	return &http.Response{StatusCode: code}, nil
+}
+
+func newDoneContext() context.Context {
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	return ctx
 }

